File: libraries/ary/ary1Cpync.c
#include "sae_par.h"
#include "ary1.h"
#include "star/hds.h"

void ary1Cpync( HDSLoc *loc1, const char *name, HDSLoc *loc2,
                int *status ) {
/*
*+
*  Name:
*     ary1Cpync

*  Purpose:
*     Copy a named HDS component from one structure to another.

*  Synopsis:
*     void ary1Cpync( HDSLoc *loc1, const char *name, HDSLoc *loc2,
*                     int *status )

*  Description:
*     This function copies a named HDS component (if it exists) from one
*     structure to another, preserving its name in the process. If the
*     component to be copied does not exist, then the routine completes
*     without action, but no error results. An error will be reported,
*     however, if a component exists to be copied but a component of
*     the same name already exists in the output structure.

*  Parameters:
*     loc1
*        Locator to input HDS structure.
*     name
*        HDS name of the component to be copied.
*     loc2
*        Locator to the HDS structure which is to receive the copied
*        component.
*     status
*        The global status.

*  Copyright:
*      Copyright (C) 2017 East Asian Observatory
*      All rights reserved.

*  Licence:
*     This program is free software; you can redistribute it and/or
*     modify it under the terms of the GNU General Public License as
*     published by the Free Software Foundation; either version 2 of
*     the License, or (at your option) any later version.
*
*     This program is distributed in the hope that it will be
*     useful,but WITHOUT ANY WARRANTY; without even the implied
*     warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR
*     PURPOSE. See the GNU General Public License for more details.
*
*     You should have received a copy of the GNU General Public License
*     along with this program; if not, write to the Free Software
*     Foundation, Inc., 51 Franklin Street,Fifth Floor, Boston, MA
*     02110-1301, USA

*  Authors:
*     RFWS: R.F. Warren-Smith (STARLINK)
*     DSB: David S. Berry (EAO)

*  History:
*     03-JUL-2017 (DSB):
*        Original version, based on equivalent Fortran routine by RFWS.

*-
*/

/* Local variables: */
   int there;                 /* Whether the component exists */
   HDSLoc *tloc = NULL;       /* Temporary locator */

/* Check inherited global status. */
   if( *status != SAI__OK ) return;

/* Initialise. */
   there = 1;

/* See if the component to be copied exists. */
   datThere( loc1, name, &there, status );
   if( ( *status == SAI__OK ) && there ){

/* If so, then locate it and copy it. */
      datFind( loc1, name, &tloc, status );
      datCopy( tloc, loc2, name, status );

/* Annul the locator when done. */
      datAnnul( &tloc, status );
   }

/* Call error tracing routine and exit. */
   if( *status != SAI__OK ) ary1Trace( "ary1Cpync", status );

}

File: libraries/fary/ary1_cpync.f
      SUBROUTINE ARY1_CPYNC( LOC1, NAME, LOC2, STATUS )
*+
*  Name:
*     ARY1_CPYNC

*  Purpose:
*     Copy a named HDS component from one structure to another.

*  Language:
*     Starlink Fortran 77

*  Invocation:
*     CALL ARY1_CPYNC( LOC1, NAME, LOC2, STATUS )

*  Description:
*     The routine copies a named HDS component (if it exists) from one
*     structure to another, preserving its name in the process. If the
*     component to be copied does not exist, then the routine completes
*     without action, but no error results. An error will be reported,
*     however, if a component exists to be copied but a component of
*     the same name already exists in the output structure.

*  Arguments:
*     LOC1 = CHARACTER * ( * ) (Given)
*        Locator to input HDS structure.
*     NAME = CHARACTER * ( * ) (Given)
*        HDS name of the component to be copied.
*     LOC2 = CHARACTER * ( * ) (Given)
*        Locator to the HDS structure which is to receive the copied
*        component.
*     STATUS = INTEGER (Given and Returned)
*        The global status.

*  Algorithm:
*     -  Initialise.
*     -  See if the component to be copied exists.
*     -  If it does, then locate it and copy it.
*     -  Annul the locator when done.

*  Authors:
*     RFWS: R.F. Warren-Smith (STARLINK)
*     {enter_new_authors_here}

*  History:
*     8-AUG-1989 (RFWS):
*        Original version.
*     {enter_changes_here}

*  Bugs:
*     {note_any_bugs_here}

*-
      
*  Type Definitions:
      IMPLICIT NONE              ! No implicit typing

*  Global Constants:
      INCLUDE 'SAE_PAR'          ! Standard SAE constants
      INCLUDE 'DAT_PAR'          ! DAT_ public constants
      INCLUDE 'ARY_CONST'        ! ARY_ private constants

*  Arguments Given:
      CHARACTER * ( * ) LOC1
      CHARACTER * ( * ) NAME
      CHARACTER * ( * ) LOC2

*  Status:
      INTEGER STATUS             ! Global status

*  Local Variables:
      LOGICAL THERE              ! Whether the component exists
      CHARACTER * ( DAT__SZLOC ) TLOC ! Temporary locator

*.

*  Check inherited global status.
      IF ( STATUS .NE. SAI__OK ) RETURN

*  Initialise.
      THERE = .TRUE.

*  See if the component to be copied exists.
      CALL DAT_THERE( LOC1, NAME, THERE, STATUS )
      IF ( ( STATUS .EQ. SAI__OK ) .AND. THERE ) THEN

*  If so, then locate it and copy it.
         TLOC = ARY__NOLOC
         CALL DAT_FIND( LOC1, NAME, TLOC, STATUS )
         CALL DAT_COPY( TLOC, LOC2, NAME, STATUS )

*  Annul the locator when done.
         CALL DAT_ANNUL( TLOC, STATUS )
         TLOC = ARY__NOLOC
      END IF
       
*  Call error tracing routine and exit.
      IF ( STATUS .NE. SAI__OK ) CALL ARY1_TRACE( 'ARY1_CPYNC', STATUS )

      END
